storage/src/nodestore/mod.rs

```diff
@@ -681,23 +617,92 @@ where
     T: Parentable,
     S: ReadableStorage,
 {
-    fn root_address(&self) -> Option<LinearAddress> {
-        self.header.root_address()
-    }
-
     fn root_hash(&self) -> Option<TrieHash> {
         self.kind.root_hash()
     }
 }
 
+impl<T, S: ReadableStorage> NodeStore<T, S> {
+    /// Read a [Node] from the provided [`LinearAddress`].
+    /// `addr` is the address of a `StoredArea` in the `ReadableStorage`.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`FileIoError`] if the node cannot be read.
+    pub fn read_node_from_disk(
+        &self,
+        addr: LinearAddress,
+        mode: &'static str,
+    ) -> Result<SharedNode, FileIoError> {
+        if let Some(node) = self.storage.read_cached_node(addr, mode) {
+            return Ok(node);
+        }
+
+        debug_assert!(addr.get() % 8 == 0);
+
+        // saturating because there is no way we can be reading at u64::MAX
+        // and this will fail very soon afterwards
+        let actual_addr = addr.get().saturating_add(1); // skip the length byte
+
+        let _span = fastrace::local::LocalSpan::enter_with_local_parent("read_and_deserialize");
+
+        let area_stream = self.storage.stream_from(actual_addr)?;
+        let node: SharedNode = Node::from_reader(area_stream)
+            .map_err(|e| {
+                self.storage
+                    .file_io_error(e, actual_addr, Some("read_node_from_disk".to_string()))
+            })?
+            .into();
+        match self.storage.cache_read_strategy() {
+            CacheReadStrategy::All => {
+                self.storage.cache_node(addr, node.clone());
+            }
+            CacheReadStrategy::BranchReads => {
+                if !node.is_leaf() {
+                    self.storage.cache_node(addr, node.clone());
+                }
+            }
+            CacheReadStrategy::WritesOnly => {}
+        }
+        Ok(node)
+    }
+
+    /// Returns (index, `area_size`) for the stored area at `addr`.
+    /// `index` is the index of `area_size` in the array of valid block sizes.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`FileIoError`] if the area cannot be read.
+    pub fn area_index_and_size(
+        &self,
+        addr: LinearAddress,
+    ) -> Result<(AreaIndex, u64), FileIoError> {
+        let mut area_stream = self.storage.stream_from(addr.get())?;
+
+        let index: AreaIndex = area_stream.read_byte().map_err(|e| {
+            self.storage.file_io_error(
+                Error::new(ErrorKind::InvalidData, e),
+                addr.get(),
+                Some("deserialize".to_string()),
```

Comment:
nit: or even more context than this

```suggestion
                Some("area_index_and_size".to_string()),
```